### PR TITLE
[AMD-AIE] Add a pass to tile and fuse TileInterface ops using scf.forall

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit adds the initial effort for tiling + fusing TileInterface ops using scf.forall by creating a new pass `--iree-amdaie-tile-and-fuse`.
-- It creates a new concise transform dialect script and adds a `--iree-amdaie-tile-and-fuse` test case to demo the generalised feature.

Co-authored-by: Nirvedh Meshram <nmeshram@amd.com>
Co-authored-by: Abhishek Varma <abhvarma@amd.com>

### CODE REVIEW

File: tests/samples/matmul_fill_spec_pad_new.mlir [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+//   ${IREE_DIR}/build/tools/iree-opt \
+//     ${IREE_AMD_AIE_DIR}/tests/samples/matmul_fill_static.mlir \
+//     --iree-hal-target-backends=amd-aie \
+//     --iree-abi-transformation-pipeline \
+//     --iree-flow-transformation-pipeline \
+//     --iree-stream-transformation-pipeline \
+//     --iree-hal-configuration-pipeline | \
+//   ${IREE_DIR}/build/tools/iree-opt \
+//      --pass-pipeline='builtin.module(hal.executable(hal.executable.variant(iree-hal-translate-target-executable-variants{target=amd-aie})))' \
+//      --iree-codegen-transform-dialect-library=${IREE_AMD_AIE_DIR}/tests/samples/matmul_fill_spec_pad.mlir
```
Comment (nirvedhmeshram) [inline tests/samples/matmul_fill_spec_pad_new.mlir:15]: This should be matmul_fill_spec_pad_new.mlir?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp [hunk 2]
```diff
@@ -21,7 +21,17 @@ namespace mlir::iree_compiler::AMDAIE {
...
 void buildAMDAIETransformPassPipeline(OpPassManager &pm) {
   addCommonTargetExecutablePreprocessingPasses(pm);
+  // TODO: Current we don't include C++ equivalent of Transform
+  //       dialect scripts. We are thus guarding their inclusion
+  //       with a bool `useCPlusPlusTransformPasses` which would
+  //       be used during development efforts. Once the C++ passes
+  //       are ready, we will include these passes by default and
+  //       take away the guarding.
+  bool useCPlusPlusTransformPasses = false;
+  if (useCPlusPlusTransformPasses)
```
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.cpp:31]: IIUC you want to set this to true when using the matmul_fill_spec_pad_new.mlir script? If thats the case I think what you want is a global option that you can set from the command as true, here is an example, 
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp#L56-L59
so then when running with pad_new.mlir you can set it to true

File: tests/samples/matmul_fill_spec_pad_new.mlir (hunk 1 above)
Comment (nirvedhmeshram) [reply tests/samples/matmul_fill_spec_pad_new.mlir:15]: You can also add `useCPlusPlusTransformPasses` as a flag mentioned in another comment

File: tests/samples/tile_and_fuse.mlir [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    } -> tensor<?x?xf32>
+  return %2 : tensor<?x?xf32>
+}
+//      CHECK: @matmul_bias_add
+//      CHECK:   scf.forall
+// CHECK-SAME:   {
+//      CHECK:       linalg.fill
+//      CHECK:       linalg.matmul
+//      CHECK:       linalg.generic
+//      CHECK:   }
```
Comment (nirvedhmeshram) [inline tests/samples/tile_and_fuse.mlir:60]: Nit: Add newline at end of file

Comment (nirvedhmeshram) [inline tests/samples/tile_and_fuse.mlir:60]: Nit: Add newline at end of file

**Review by nirvedhmeshram — COMMENTED** [review]
I will let Mahesh review on the core logic, but some minor comments.

**Review by yzhang93 — COMMENTED** [review]
I haven't compared this to the original pad transform dialect script. Could you summarize what has been changed?

File: tests/samples/matmul_fill_spec_pad_new.mlir [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      transform.apply_patterns.iree.fold_fill_into_pad
+      transform.apply_patterns.scf.for_loop_canonicalization
+      transform.apply_patterns.canonicalization
+    } : !transform.any_op
+    transform.iree.apply_licm %func : !transform.any_op
+    transform.iree.apply_cse %func : !transform.any_op
+    transform.yield
+  }
+
+  transform.named_sequence @__transform_main(%variant_op: !transform.any_op {transform.read_only}) {
```
Comment (yzhang93) [inline tests/samples/matmul_fill_spec_pad_new.mlir:32]: I haven't compared this to the original pad transform dialect script. Could you summarize what has been changed?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  LLVM_DEBUG(llvm::dbgs() << "consumerOp: " << consumerOp << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "tilingLevel: " << tilingLevel << "\n");
+  // TODO: Currently hardcoding the tile size to make the tile and fuse run for
+  //       at least the first level. So, we can generalize this by adding tile
+  //       level, similar to LLVMCPUTileAndFuse.cpp.
+  SmallVector<OpFoldResult> tileSizes = getAsIndexOpFoldResult(context, {8, 8});
+  auto options = scf::SCFTilingOptions().setTileSizes(tileSizes);
+  options.setMapping(
+      {gpu::GPUBlockMappingAttr::get(context, gpu::MappingId::DimY),
+       gpu::GPUBlockMappingAttr::get(context, gpu::MappingId::DimX)});
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:320]: This might be ok for now, but ideally we'll need some AIE specific attributes similar to the gpu ones.

**Comment by Abhishek-Varma** [comment]
> I haven't compared this to the original pad transform dialect script. Could you summarize what has been changed?

Sure, the change is only the [first level of tiling and fusing](https://github.com/nod-ai/iree-amd-aie/blob/main/tests/samples/matmul_fill_spec_pad.mlir#L36-L44) where the tile_size=8x8.
In the new transform dialect script we were able to get rid of these exact lines and in turn depend on the C++ pass.

File: tests/samples/tile_and_fuse.mlir [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        linalg.yield %3 : f32
+    } -> tensor<?x?xf32>
+  return %2 : tensor<?x?xf32>
+}
+//      CHECK: @matmul_bias_add
+//      CHECK:   scf.forall
+// CHECK-SAME:   {
+//      CHECK:       linalg.fill
+//      CHECK:       linalg.matmul
+//      CHECK:       linalg.generic
```
Comment (yzhang93) [inline tests/samples/tile_and_fuse.mlir:59]: What is the specific part of this test? Could you elaborate the result IR?

Comment (Abhishek-Varma) [reply tests/samples/tile_and_fuse.mlir:59]: I didn't understand. Do you mean to suggest to have the entire result IR captured in the lit test? I don't think that should be necessary though.

I have taken the base test from [iree/LLVMCPU/tile_and_fuse.mlir](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse.mlir#L25-L34) and they too aren't capturing the entire resulting IR.

This is just to demonstrate that indeed the pass will take up the last TileInterface operation in the IR, and start tiling+fusing and will indeed just have one scf.forall (I can perhaps add a stricter CHECK-NEXT statements to ensure that ending with the return statement).

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp (hunk 5 above)
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:320]: Yep, you're correct. We'll be revisiting this all iteratively. Currently we're trying to mimic exactly what the [pad transform dialect script gets us](https://github.com/nod-ai/iree-amd-aie/blob/1e2aa24c856ce33fcef3d749e98c4ac2aff407ac/tests/samples/matmul_fill_spec_pad.mlir#L39). Similar thing for the pack transform dialect script, for now.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td [hunk 7]
```diff
@@ -20,4 +20,14 @@ def AMDAIELowerExecutableTarget :
   let constructor = "mlir::iree_compiler::AMDAIE::createAMDAIELowerExecutableTargetPass()";
 }
 
+def AMDAIETileAndFuse :
+    Pass<"iree-amdaie-tile-and-fuse", ""> {
+  let summary = "Pass to tile and fuse TilingInterface operations.";
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td:25]: This summary doesn't really explain what the pass does.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+  }
+  return scf::SCFFuseProducerOfSliceResult{fusableProducer,
+                                           tileAndFuseResult->tiledValues[0],
+                                           tileAndFuseResult->tiledOps};
+}
+
+/// Starting from `op` walk all operands backwards to find all
+/// potentially fusable operations, i.e. operations that implement
+/// the `TilingInterface`.
+static void collectTiledAndFusedOps(Operation *rootOp,
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:131]: I hope the plan is to re-use the code in IREE, to avoid code duplication.

File: tests/samples/tile_and_fuse.mlir (hunk 6 above)
Comment (nirvedhmeshram) [reply tests/samples/tile_and_fuse.mlir:76]: Yes actually we want to keep the CHECK's general enough so that they dont break with minor changes like ops moving around.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp (hunk 8 above)
Comment (nirvedhmeshram) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:179]: Yes this will be refactored once Mahesh finishes upstream changes to allow scf for_all tiling

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td (hunk 7 above)
Comment (nirvedhmeshram) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Passes.td:25]: This is lifted from here
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.td#L128
These summary's are high level mentioning what it accomplishes which in this case is tile and fuse, how its done should be explained in the cpp file doing it.

File: tests/samples/matmul_fill_spec_pad_new.mlir (hunk 4 above)
Comment (nirvedhmeshram) [reply tests/samples/matmul_fill_spec_pad_new.mlir:35]: added a comment explaining that

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+    // operation, do nothing. do nothing.
+    return success();
+  }
+  if (!llvm::hasSingleElement(countOps)) {
+    return exportOp->emitOpError(
+        "unexpected multiple flow.dispatch.workgroup_count_from_dag_root "
+        "operations "
+        "in body");
+  }
+  return TypeSwitch<Operation *, LogicalResult>(countOps[0])
```
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:69]: Instead of this for now you can just replace the operation with all results being value 1 for now.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+        return lowerWorkgroupCountFromSliceOp(rewriter, countOp, entryPointFn,
+                                              workgroupCount,
+                                              maxWorkgroupParallelDims);
+      })
+      .Default([&](Operation *) { return success(); });
+}
+
+/// Clones the operation and updates the destination if the operation
+/// implements the `DestinationStyleOpInterface`.
+static Operation *cloneOpAndUpdateDestinationArgs(RewriterBase &rewriter,
```
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:80]: For all methods that care copied over from core can you just demaracte them with 
```
//===----------------------------------------------------------------------------------===//
// Methods copied over from core to implement tile and fuse with scf.forall.
// TODO(MaheshRavishankar) Replace them with core methods when upstream
// can handle this natively
//===----------------------------------------------------------------------------------===//
```

and

//===----------------------------------------------------------------------------------===//
// End methods copied over from core to implement tile and fuse with scf.forall.
//===----------------------------------------------------------------------------------===//
```

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice! Thanks @Abhishek-Varma and @nirvedhmeshram . I will take the AI of replacing it with core methods when they are ready (maybe a week or so away). Just a few minor changes.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp (hunk 9 above)
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:69]: I didn't understand this point.

Can you please elaborate? Also, the intention behind the same.

Comment (nirvedhmeshram) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:69]: I can add this, @Abhishek-Varma the intention is that the workgroup count lowering is dummy for the aie backend as we dont have this concept in our programming model, so this logic that I blindly lefted from the cpu backend can be greatly simplified.

**Comment by newling** [comment]
@MaheshRavishankar  Is there a reason bufferization done in the transform dialect (matmul_fill_spec_pad.mlir)? Can it be done as a separate pass in Passes.cpp?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+  }
+  auto forallLoop = cast<scf::ForallOp>(tilingResult->loops[0]);
+  yieldedValuesToOrigValues.append(rootOp->result_begin(),
+                                   rootOp->result_end());
+  // A map from untiled value to scf.forall shared_outs. The shared_outs is used
+  // for DPS init operand if they use the same init operands.
+  llvm::DenseMap<Value, Value> mapToSharedOuts;
+
+  // WAR for `if` ops generating `scf.if` operations.
+  if (auto rootPadOp = dyn_cast<tensor::PadOp>(rootOp)) {
```
Comment (MaheshRavishankar) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp:258]: The pad stuff seems like a distraction for now... If you want to remove it that would be fine. We might need it later if we want to handle pad, but we can punt on it for now.

**Review by MaheshRavishankar — APPROVED** [review]
